src/D2L.Security.OAuth2/Keys/KeyManagementService.cs

```diff
@@ -0,0 +1,226 @@
+				PersistKeyInCsp = false
+			};
+
+			var pub = csp.ExportParameters(
+				includePrivateParameters: false
+			);
+
+			var priv = csp.ExportCspBlob(
+				includePrivateParameters: true
+			);
+
+			var expiresAt = now + m_config.KeyLifetime;
+
+			await Task.WhenAll(
+				m_publicKeys.SaveAsync(
+					keyId,
+					new RsaJsonWebKey(
+						keyIdStr,
+						expiresAt,
+						pub
+					)
+				),
+				m_privateKeys.SaveAsync(
+					new PrivateKeyData(
+						id: keyIdStr,
+						kind: PrivateKeyData.KeyKinds.Rsa,
+						data: priv,
+						createdAt: now,
+						notBefore: now + m_config.KeyTimeUntilUse,
+						expiresAt: expiresAt
+					)
+				)
+			).ConfigureAwait( false );
+		}
+
+		private async Task RefreshKeyAsync( DateTimeOffset now ) {
+			var keys = await m_privateKeys.GetAllAsync(
+				validUntilAtLeast: now
+			).ConfigureAwait( false );
+
+			var best = ChooseKey( keys, now )?.Ref();
+
+			if( best == null ) {
+				// If we didn't find anything in the database, continue with
+				// the one we have (if any)
+				return;
+			}
+
+			var prev = Interlocked.Exchange( ref m_current, best );
+
+			prev?.Dispose();
+		}
+
+		private D2LSecurityToken ChooseKey(
+			IEnumerable<PrivateKeyData> keys,
+			DateTimeOffset now
+		) {
+			PrivateKeyData candidate = null;
+
+			foreach( var key in keys ) {
+				// Ignore unsupported key types
+				if( key.Kind != PrivateKeyData.KeyKinds.Rsa ) {
+					continue;
+				}
+
+				// The data provider should filter expired keys, but we don't
+				// rely on that just in case.
+				if( key.IsExpired( now ) ) {
+					continue;
+				}
+
+				// Use any non-expired key if it's the only one
+				if( candidate == null ) {
+					candidate = key;
+					continue;
+				}
+
+				var candidateIsPastNotBefore = candidate.IsPastNotBefore( now );
+				var keyIsPastNotBefore = key.IsPastNotBefore( now );
+
+				if( !candidateIsPastNotBefore && keyIsPastNotBefore ) {
+					// Prefer keys past their "NotBefore" date because they are
+					// guarunteed to not be excluded from caches.
+					candidate = key;
+				} else if( !candidateIsPastNotBefore && key.CreatedAt < candidate.CreatedAt ) {
+					// When comparing two keys that are not past their NotBefore
+					// points, prefer the oldest one (same cache rationale.)
+					candidate = key;
+				} else if( candidateIsPastNotBefore && keyIsPastNotBefore && candidate.CreatedAt < key.CreatedAt ) {
+					// If we have two keys past their NotBefore date, pick the
+					// one that was created most recently.
+					candidate = key;
+				}
+			}
+
+			if( candidate == null ) {
+				return null;
+			}
+
+			Func<Tuple<AsymmetricSecurityKey, IDisposable>> keyFactory = candidate.Kind switch {
+				PrivateKeyData.KeyKinds.Rsa => () => {
+					var csp = new RSACryptoServiceProvider {
+						PersistKeyInCsp = false
+					};
+
+					csp.ImportCspBlob( candidate.Data );
+
+					var key = new RsaSecurityKey( csp );
+
+					return new(key, csp);
+				},

```

Comment:
Believe you had ecdsa as a TODO. Do you want me to fully review what's in place now, and then that'll get added before landing?

Reply:
🤷 I was going to do it as a follow-up.
(It isn't something we currently use -- we do use the validation side of things, but we never generate those sort of keys.)

Reply:
(Supporting it here isn't tricky, it just seemed like a distraction from this PR.)